utils/treesitter/grammar.js

```diff
@@ -0,0 +1,554 @@
+        [PREC.Relational, choice('==', '!=', '<', '<=', '>', '>=')],
+        [PREC.Additive, choice('+', '-')],
+        [PREC.Multiplicative, choice($.binary_star, '/', '%')],
+      ];
+
+      return choice(
+        ...table.map(([precedence, operator]) =>
+          prec.left(
+            precedence,
+            seq(
+              field('left', $._expression),
+              field('operator', operator),
+              field('right', $._expression)
+            )
+          )
+        )
+      );
+    },
+
+    // This should be non-associative but conflicts are not allowed in tree-sitter
+    as_expression: ($) => prec.left(seq($._expression, 'as', $._expression)),
+
+    ref_expression: ($) => prec.right(PREC.TermPrefix, seq('&', $._expression)),
+
+    deref_expression: ($) =>
+      prec.right(PREC.TermPrefix, seq('*', $._expression)),
+
+    fn_type_expression: ($) =>
+      prec.left(seq('__Fn', $.paren_expression, '->', $._expression)),
+
+    if_expression: ($) =>
+      prec(
+        PREC.IfExpression,
+        seq('if', $._expression, 'then', $._expression, 'else', $._expression)
+      ),
+
+    paren_expression: ($) => seq('(', comma_sep($._expression), ')'),
+
+    index_expression: ($) =>
+      prec(PREC.TermPostfix, seq($._expression, '[', $._expression, ']')),
+
+    designator: ($) => seq('.', choice('base', $.ident)),
+
+    postfix_expression: ($) =>
+      prec(
+        PREC.TermPostfix,
+        seq(
+          $._expression,
+          choice(
+            '++',
+            '--',
+            $.designator,
+            seq('->', $.ident),
+            seq(choice('.', '->'), '(', $._expression, ')')
+          )
+        )
+      ),
+
+    where_clause: ($) =>
+      prec(
+        PREC.WhereClause,
+        choice(
+          seq($._expression, '==', $._expression),
+          seq($._expression, 'impls', $._expression),
+          seq($._expression, '=', $._expression),
+          // FIXME: logical and conflicts with this
+          prec.left(seq($.where_clause, 'and', $.where_clause))
+        )
+      ),
+
+    where_expression: ($) =>
+      prec.left(PREC.TermPostfix, seq($._expression, 'where', $.where_clause)),
+
+    call_expression: ($) =>
+      prec(PREC.TermPostfix, seq($._expression, $.paren_expression)),
+
+    pointer_expression: ($) =>
+      prec(PREC.TypePostfix, seq($._expression, $.postfix_star)),
+
+    _expression: ($) =>
+      choice(
+        $.array_literal,
+        $.as_expression,
+        $.binary_expression,
+        $.builtin_type,
+        $.call_expression,
+        $.deref_expression,
+        $.fn_type_expression,
+        $.ident,
+        $.if_expression,
+        $.index_expression,
+        $.literal,
+        $.paren_expression,
+        $.pointer_expression,
+        $.postfix_expression,
+        $.ref_expression,
+        $.unary_prefix_expression,
+        $.where_expression,
+        'self',
+        '.Self',
+        $.designator
```

Comment:
These were added back again. These are used in where_clause's expression

Reply:
Will they be removed once where clauses get their own rules instead of reusing expression?

Reply:
```suggestion
        // TODO: Remove these two once `where` clauses don't use the expression rule
        '.Self',
        $.designator
```